Go client Reader panics when passing custom StartMessageID

#### Expected behavior

This is refers to the same issue in cgo based Go client https://github.com/apache/pulsar/issues/3180, which is applicable here as well.

Reader client's [CreateReader](https://github.com/apache/pulsar-client-go/blob/master/pulsar/client.go#L109) API should not panic at runtime when using a custom type satisfying the [MessageID](https://github.com/apache/pulsar-client-go/blob/master/pulsar/message.go#L108) interface , in [ReaderOptions](https://github.com/apache/pulsar-client-go/blob/master/pulsar/reader.go#L48) argument.

e.g. Custom type
```go
type myMessageID struct {
	data []byte
}

func (id *myMessageID) Serialize() []byte {
	return id.data
}
```

e.g. Usage snippet (compiles)
```go

// send 10 messages
msgIDs := [10]MessageID{}
for i := 0; i < 10; i++ {
	msgID, err := producer.Send(ctx, &ProducerMessage{
		Payload: []byte(fmt.Sprintf("hello-%d", i)),
	})
	assert.NoError(t, err)
	assert.NotNil(t, msgID)
	msgIDs[i] = msgID
}

// custom start message ID
// I'm referencing just the []byte data from some test data slice
myStartMsgId := &myMessageID{
	data: msgIDs[4].Serialize(),
}

reader, err = client.CreateReader(ReaderOptions{
	Topic:          topic,
	StartMessageID: myStartMsgId,
})
```


#### Actual behavior

Panic at runtime when invoking the API
```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]                                                                                               
        panic: runtime error: invalid memory address or nil pointer dereference                                                                                                   
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xb9e8f2]                                                                                                            
                                                                                                                                                                                  
goroutine 20 [running]:                                                                                                                                                           
testing.tRunner.func1(0xc000124200)                                                                                                                                               
        /usr/local/go/src/testing/testing.go:830 +0x69d                                                                                                                           
panic(0xcf0820, 0x13abde0)                                                                                                                                                        
        /usr/local/go/src/runtime/panic.go:522 +0x1b5                                                                                                                             
github.com/apache/pulsar-client-go/pulsar.TestReaderWithCustomMessageID(0xc000124200)                                                                                             
        /pulsar-client-go/pulsar/reader_test.go:418 +0x9c2                                                                                                                        
testing.tRunner(0xc000124200, 0xda0408)                                                                                                                                           
        /usr/local/go/src/testing/testing.go:865 +0x164                                                                                                                           
created by testing.(*T).Run                                                                                                                                                       
        /usr/local/go/src/testing/testing.go:916 +0x65b    
```


#### Steps to reproduce

Here's a unit test I added locally to reproduce it.
[reader_panic_test.diff.txt](https://github.com/apache/pulsar-client-go/files/4874522/reader_panic_test.diff.txt)

Output snippet for test when running the client tests in test container
```bash
./docker-ci.sh
```

```bash
+ go test -race -coverprofile=/tmp/coverage -timeout=20m ./... -run=TestReaderWithCustomMessageID                                                                                 
?       github.com/apache/pulsar-client-go/examples/consumer    [no test files]                                                                                                   
?       github.com/apache/pulsar-client-go/examples/consumer-listener   [no test files]                                                                                           
?       github.com/apache/pulsar-client-go/examples/producer    [no test files]                                                                                                   
?       github.com/apache/pulsar-client-go/examples/reader      [no test files]                                                                                                   
ok      github.com/apache/pulsar-client-go/integration-tests    1.022s  coverage: 0.0% of statements [no tests to run]                                                            
?       github.com/apache/pulsar-client-go/perf [no test files]                                                                                                                   
time="2020-07-05T05:54:05Z" level=info msg="Connecting to broker" remote_addr="pulsar://localhost:6650"                                                                           
time="2020-07-05T05:54:05Z" level=info msg="TCP connection established" local_addr="127.0.0.1:49144" remote_addr="pulsar://localhost:6650"                                        
time="2020-07-05T05:54:05Z" level=info msg="Connection is ready" local_addr="127.0.0.1:49144" remote_addr="pulsar://localhost:6650"                                               
time="2020-07-05T05:54:06Z" level=info msg="Created producer" cnx="127.0.0.1:49144 -> 127.0.0.1:6650" producer_name=standalone-0-0 topic="persistent://public/default/my-topic-501
323100"                                                                                                                                                                           
time="2020-07-05T05:54:06Z" level=info msg="Closing producer" producer_name=standalone-0-0 topic="persistent://public/default/my-topic-501323100"                                 
time="2020-07-05T05:54:06Z" level=info msg="Closed producer" producer_name=standalone-0-0 topic="persistent://public/default/my-topic-501323100"                                  
--- FAIL: TestReaderWithCustomMessageID (1.14s)                                                                                                                                   
    reader_test.go:410:                                                                                                                                                           
                Error Trace:    reader_test.go:410                                                                                                                                
                Error:          func (assert.PanicTestFunc)(0xba3710) should not panic                                                                                            
                                        Panic value:    interface conversion: pulsar.MessageID is *pulsar.myMessageID, not *pulsar.messageID                                      
                Test:           TestReaderWithCustomMessageID                                                                                                                     
panic: runtime error: invalid memory address or nil pointer dereference [recovered]                                                                                               
        panic: runtime error: invalid memory address or nil pointer dereference                                                                                                   
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xb9e8f2]                                                                                                            
                                                                                                                                                                                  
goroutine 20 [running]:                                                                                                                                                           
testing.tRunner.func1(0xc000124200)                                                                                                                                               
        /usr/local/go/src/testing/testing.go:830 +0x69d                                                                                                                           
panic(0xcf0820, 0x13abde0)                                                                                                                                                        
        /usr/local/go/src/runtime/panic.go:522 +0x1b5                                                                                                                             
github.com/apache/pulsar-client-go/pulsar.TestReaderWithCustomMessageID(0xc000124200)                                                                                             
        /pulsar-client-go/pulsar/reader_test.go:418 +0x9c2                                                                                                                        
testing.tRunner(0xc000124200, 0xda0408)                                                                                                                                           
        /usr/local/go/src/testing/testing.go:865 +0x164                                                                                                                           
created by testing.(*T).Run                                                                                                                                                       
        /usr/local/go/src/testing/testing.go:916 +0x65b                                                                                                                           
FAIL    github.com/apache/pulsar-client-go/pulsar       1.164s                                                                                                                    
ok      github.com/apache/pulsar-client-go/pulsar/internal      1.029s  coverage: 0.0% of statements [no tests to run]                                                            
ok      github.com/apache/pulsar-client-go/pulsar/internal/auth 1.021s  coverage: 0.0% of statements [no tests to run]                                                            
ok      github.com/apache/pulsar-client-go/pulsar/internal/compression  1.025s  coverage: 17.9% of statements [no tests to run]                                                   
?       github.com/apache/pulsar-client-go/pulsar/internal/pulsar_proto [no test files]
```

#### System configuration
**Pulsar version**: 2.6.0

--- a/pulsar/reader_impl.go
+++ b/pulsar/reader_impl.go
@@ -44,6 +44,19 @@ func newReader(client *client, options ReaderOptions) (Reader, error) {
 		return nil, newError(ResultInvalidConfiguration, "StartMessageID is required")
 	}
 
+	var startMessageID *messageID
+	var ok bool
+	if startMessageID, ok = options.StartMessageID.(*messageID); !ok {
+		// a custom type satisfying MessageID may not be a *messageID
+		// so re-create *messageID using its data
+		deserMsgID, err := deserializeMessageID(options.StartMessageID.Serialize())
+		if err != nil {
+			return nil, err
+		}
+		// de-serialized MessageID is a *messageID
+		startMessageID = deserMsgID.(*messageID)
+	}
+
 	subscriptionName := options.SubscriptionRolePrefix
 	if subscriptionName == "" {
 		subscriptionName = "reader"
@@ -61,7 +74,7 @@ func newReader(client *client, options ReaderOptions) (Reader, error) {
 		subscription:               subscriptionName,
 		subscriptionType:           Exclusive,
 		receiverQueueSize:          receiverQueueSize,
-		startMessageID:             options.StartMessageID.(*messageID),
+		startMessageID:             startMessageID,
 		startMessageIDInclusive:    options.StartMessageIDInclusive,
 		subscriptionMode:           nonDurable,
 		readCompacted:              options.ReadCompacted,
@@ -80,8 +93,8 @@ func newReader(client *client, options ReaderOptions) (Reader, error) {
 	if err != nil {
 		return nil, err
 	}
-	pc, err := newPartitionConsumer(nil, client, consumerOptions, reader.messageCh, dlq)
 
+	pc, err := newPartitionConsumer(nil, client, consumerOptions, reader.messageCh, dlq)
 	if err != nil {
 		close(reader.messageCh)
 		return nil, err

--- a/pulsar/reader_test.go
+++ b/pulsar/reader_test.go
@@ -402,7 +402,7 @@ func TestReaderOnSpecificMessageWithCustomMessageID(t *testing.T) {
 	}
 
 	// custom start message ID
-	myStartMsgId := &myMessageID{
+	myStartMsgID := &myMessageID{
 		data: msgIDs[4].Serialize(),
 	}
 
@@ -411,7 +411,7 @@ func TestReaderOnSpecificMessageWithCustomMessageID(t *testing.T) {
 	assert.NotPanics(t, func() {
 		reader, err = client.CreateReader(ReaderOptions{
 			Topic:          topic,
-			StartMessageID: myStartMsgId,
+			StartMessageID: myStartMsgID,
 		})
 	})
 
@@ -430,7 +430,7 @@ func TestReaderOnSpecificMessageWithCustomMessageID(t *testing.T) {
 	// create reader on 5th message (included)
 	readerInclusive, err := client.CreateReader(ReaderOptions{
 		Topic:                   topic,
-		StartMessageID:          myStartMsgId,
+		StartMessageID:          myStartMsgID,
 		StartMessageIDInclusive: true,
 	})
 

--- a/pulsar/reader_test.go
+++ b/pulsar/reader_test.go
@@ -362,3 +362,87 @@ func TestReaderHasNext(t *testing.T) {
 
 	assert.Equal(t, 10, i)
 }
+
+type myMessageID struct {
+	data []byte
+}
+
+func (id *myMessageID) Serialize() []byte {
+	return id.data
+}
+
+func TestReaderOnSpecificMessageWithCustomMessageID(t *testing.T) {
+	client, err := NewClient(ClientOptions{
+		URL: lookupURL,
+	})
+
+	assert.Nil(t, err)
+	defer client.Close()
+
+	topic := newTopicName()
+	ctx := context.Background()
+
+	// create producer
+	producer, err := client.CreateProducer(ProducerOptions{
+		Topic:           topic,
+		DisableBatching: true,
+	})
+	assert.Nil(t, err)
+	defer producer.Close()
+
+	// send 10 messages
+	msgIDs := [10]MessageID{}
+	for i := 0; i < 10; i++ {
+		msgID, err := producer.Send(ctx, &ProducerMessage{
+			Payload: []byte(fmt.Sprintf("hello-%d", i)),
+		})
+		assert.NoError(t, err)
+		assert.NotNil(t, msgID)
+		msgIDs[i] = msgID
+	}
+
+	// custom start message ID
+	myStartMsgId := &myMessageID{
+		data: msgIDs[4].Serialize(),
+	}
+
+	// attempt to create reader on 5th message (not included)
+	var reader Reader
+	assert.NotPanics(t, func() {
+		reader, err = client.CreateReader(ReaderOptions{
+			Topic:          topic,
+			StartMessageID: myStartMsgId,
+		})
+	})
+
+	assert.Nil(t, err)
+	defer reader.Close()
+
+	// receive the remaining 5 messages
+	for i := 5; i < 10; i++ {
+		msg, err := reader.Next(context.Background())
+		assert.NoError(t, err)
+
+		expectMsg := fmt.Sprintf("hello-%d", i)
+		assert.Equal(t, []byte(expectMsg), msg.Payload())
+	}
+
+	// create reader on 5th message (included)
+	readerInclusive, err := client.CreateReader(ReaderOptions{
+		Topic:                   topic,
+		StartMessageID:          myStartMsgId,
+		StartMessageIDInclusive: true,
+	})
+
+	assert.Nil(t, err)
+	defer readerInclusive.Close()
+
+	// receive the remaining 6 messages
+	for i := 4; i < 10; i++ {
+		msg, err := readerInclusive.Next(context.Background())
+		assert.NoError(t, err)
+
+		expectMsg := fmt.Sprintf("hello-%d", i)
+		assert.Equal(t, []byte(expectMsg), msg.Payload())
+	}
+}

